app/engine/Flywheel.js

```diff
@@ -0,0 +1,337 @@
+  are part of a specific phase, i.e. dirtyDataPoints[flankLength], which will be eliminated from the flank
+
+  The calculation of angular velocity and acceleration is based on Quadratic Regression, as the second derivative tends to be
+  quite fragile when small errors are thrown in the mix. The math behind this approach can be found in https://physics.info/motion-equations/
+  which is intended for simple linear motion, but the formula are identical when applied to angular distances, velocities and
+  accelerations.
+*/
+
+import loglevel from 'loglevel'
+import { createStreamFilter } from './utils/StreamFilter.js'
+import { createSeries } from './utils/Series.js'
+import { createOLSLinearSeries } from './utils/OLSLinearSeries.js'
+import { createTSQuadraticSeries } from './utils/FullTSQuadraticSeries.js'
+const log = loglevel.getLogger('RowingEngine')
+
+function createFlywheel (rowerSettings) {
+  const angularDisplacementPerImpulse = (2.0 * Math.PI) / rowerSettings.numOfImpulsesPerRevolution
+  const flankLength = Math.max(3, rowerSettings.flankLength)
+  const minimumDragFactorSamples = Math.floor(rowerSettings.minimumRecoveryTime / rowerSettings.maximumTimeBetweenImpulses)
+  const minumumTorqueBeforeStroke = rowerSettings.minumumForceBeforeStroke * (rowerSettings.sprocketRadius / 100)
+  const currentDt = createStreamFilter(rowerSettings.smoothing, rowerSettings.maximumTimeBetweenImpulses)
+  const _deltaTime = createOLSLinearSeries(flankLength)
+  const _angularDistance = createTSQuadraticSeries(flankLength)
+  const _angularVelocityMatrix = []
+  const _angularAccelerationMatrix = []
+  const drag = createStreamFilter(rowerSettings.dragFactorSmoothing, (rowerSettings.dragFactor / 1000000))
+  const recoveryDeltaTime = createOLSLinearSeries()
+  const strokedetectionMinimalGoodnessOfFit = rowerSettings.minimumStrokeQuality
+  const minumumRecoverySlope = createStreamFilter(rowerSettings.dragFactorSmoothing, rowerSettings.minumumRecoverySlope)
+  let _deltaTimeBeforeFlank
+  let _angularVelocityAtBeginFlank
+  let _angularVelocityBeforeFlank
+  let _angularAccelerationAtBeginFlank
+  let _angularAccelerationBeforeFlank
+  let _torqueAtBeginFlank
+  let _torqueBeforeFlank
+  let lastKnownGoodDatapoint
+  let inRecoveryPhase
+  let maintainMetrics
+  let totalNumberOfImpulses
+  let totalTimeSpinning
+  let currentCleanTime
+  let currentRawTime
+  let currentAngularDistance
+  reset()
+  maintainMetrics = true
+
+  function pushValue (dataPoint) {
+    if (dataPoint <= rowerSettings.minimumTimeBetweenImpulses || dataPoint > (1.7 * lastKnownGoodDatapoint)) {
+      // Remove the completely deviating currentDt's, which typically are found after a long pause
+      // TODO: Omgaan met waarden boven 2 * rowerSettings.maximumTimeBetweenImpulses
+      // TODO: Omgaan met waarden boven 2 * vorige waarde --> extra impuls er tussen!
+      dataPoint = 1.7 * lastKnownGoodDatapoint
```

Comment:
> Do I understand correctly that if the dataPoint is an outlier you disregard the actual value and swap it out to a theoretical point that is the last good datapoint times a random (1.7 in the current case) value? Is this actually feasible?

No, normally the Quadratic regressor will ignore the usual outliers. However, there is one type that is problematic, that is the missed detection. In practice the time is right (usually around double the last CurrentDt), but the angular distance is missing one detection. That specific type of noise is problematic as it tends to throw off stroke detection etc.. However, since I started using the PREEMPT-kernel, I haven't seen this issue on my machine.

> I mean if we have a lot of outliers (doubnces from a reed switch) than eventually the real datapoints will be lost and we derive to a calculated set that has limited correlation to the actual curve. Below is a raw data recorded by the micro controller:

I think debounce is a very specific issue that can't be handled by a general noise filter, as it tends to be systematic, biassed (i.e. always leads to too short times) and above all tends to generate around 50% noise. As our Theil-Senn algorithm can handle "only" 39% signal to noise ratio, it will certainly fail in handling this.

I don't know if debounce is a generic issue with Reed switches. The standard NordicTrack RX-800 reed switch I encountered did not have this issue. But those guys did some pretty clever engineering where to put it mechanically, making sure vibrations in the frame were not easily going to affect it. The user manual even included a procedure to fix it when it did occur (a sensor allignment procedure). Some issues can easily be fxed by replacing the Reed switch by a better one (see for example https://github.com/laberning/openrowingmonitor/discussions/83), but that might be my limited experience with low voltage electronics speaking here.

> **My recommendation would be if the minimum rotation delta is not met (i.e. we are into a potential debounce situation), the time should be added to the next datapoint delta time (of course its not possible to determine if the the debounce belongs to the release of the reed or the activation so there is a simplification here, but I can show it with data that this does not result in material errors).**

That would be a solution, but I don't know whether that will work in all situations. I'm not 100% happy with that error handling as it is, but I noticed that the need to fix any issues in that area dissappeared when a more low-latency-oriented  oriented kernel combined with Theil-Senn regression made all my issues go away :).

Reply:
> > Do I understand correctly that if the dataPoint is an outlier you disregard the actual value and swap it out to a theoretical point that is the last good datapoint times a random (1.7 in the current case) value? Is this actually feasible?
> 
> No, normally the Quadratic regressor will ignore the usual outliers. However, there is one type that is problematic, that is the missed detection. In practice the time is right (usually around double the last CurrentDt), but the angular distance is missing one detection. That specific type of noise is problematic as it tends to throw off stroke detection etc.. However, since I started using the PREEMPT-kernel, I haven't seen this issue on my machine.
> 
> So for mine (and I gues most) rower, it behaves like dead code that will only kill the first currentDt's: in the start-up, the first CurrentDt is usually in minutes, throwing off many metrics. As there are situations with stops and pauses this will happen, this code is there to prevent this to kill metrics.
> 
> > I mean if we have a lot of outliers (doubnces from a reed switch) than eventually the real datapoints will be lost and we derive to a calculated set that has limited correlation to the actual curve. Below is a raw data recorded by the micro controller:
> 
> I think debounce is a very specific issue that can't be handled by a general noise filter, as it tends to be systematic, biassed (i.e. always leads to too short times) and above all tends to generate around 50% noise. As our Theil-Senn algorithm can handle "only" 39% signal to noise ratio, it will certainly fail in handling this.

Debounce is a very general issue. Even push buttons have debounce. There are two options to over come it: one is hardware and the other is software. The hardware requires the right combination of a capacitor and a resistor. So in a home environment, software is much more flexible and easier to implement.  

Its pretty easy to handle in code as you simply disregard any impulse that is shorter than a certain time period (e.g. the minimumTimeBetweenImpulses) and you effectively count that time to the next pulse.

> 
> I don't know if debounce is a generic issue with Reed switches. The standard NordicTrack RX-800 reed switch I encountered did not have this issue. But those guys did some pretty clever engineering where to put it mechanically, making sure vibrations in the frame were not easily going to affect it. The user manual even included a procedure to fix it when it did occur (a sensor allignment procedure). Some issues can easily be fxed by replacing the Reed switch by a better one (see for example #83), but that might be my limited experience with low voltage electronics speaking here.
> 

I would assume that they used a good hardware debounce. Vibration should not be an issue, every reed switch has a rating for the tolerance of that.

> > **My recommendation would be if the minimum rotation delta is not met (i.e. we are into a potential debounce situation), the time should be added to the next datapoint delta time (of course its not possible to determine if the the debounce belongs to the release of the reed or the activation so there is a simplification here, but I can show it with data that this does not result in material errors).**
> 
> That would be a solution, but I don't know whether that will work in all situations. I'm not 100% happy with that error handling as it is, but I noticed that the need to fix any issues in that area dissappeared when a more low-latency-oriented oriented kernel combined with Theil-Senn regression made all my issues go away :).

That is an interesting statement. I would expect a better interrupt to notice more debounce as its more sensitive. I did a short research on the Gpio libraries and I noticed this one: https://github.com/fivdi/pigpio. It claims low latency "Handle up to 20000 interrupts per second". So I tried it out and I was able to produce the debounces on the Rpi as well. Here you can find the data recorded with ORM.
[2022-09-18_21-01-58_raw.csv](https://github.com/laberning/openrowingmonitor/files/9595247/2022-09-18_21-01-58_raw.csv)

Such noise is not the singe of error but rather the sign of accuracy in my view as it proves that it can handle signals on a higher frequency, hence with better accuracy. I would consider switching to it and implement a one stage doubnce filter. Something like this (note this may not the most efficient code below :)):

```js
     if (hrDelta[0] + hrDelta[1] / 1e9 > config.rowerSettings.minimumTimeBetweenImpulses){
        hrStartTime = process.hrtime()
        const delta = hrDelta[0] + hrDelta[1] / 1e9
        process.send(delta)
      }
```

Reply:
> Its pretty easy to handle in code as you simply disregard any impulse that is shorter than a certain time period (e.g. the minimumTimeBetweenImpulses) and you effectively count that time to the next pulse.

Yes and no. On a local local level on `flywheel.js`, you could indeed ignore the impulse and add it to a debounce buffer to be added to the next impulse. However, `Rower.js` will inspect `flywheel.js` after adding a datapoint, just to find absolutely no changes are made (so it's state and metrics all will remain identical). After that, `RowingStatistics.js` will inspect `Rower.js`, also to find nothing has changed. There are some cases (for example the power/force/speed-curves in the drive phase) where this will lead to additional data that shouldn't be there. Not a big issue if it happens only once or twice in a session, but it is problematic when it happens often.

Filtering on a change in absolute time before entering any metric/state update could solve this, but it might make a lot of things much more complex.

> That is an interesting statement. I would expect a better interrupt to notice more debounce as its more sensitive. I did a short research on the Gpio libraries and I noticed this one: https://github.com/fivdi/pigpio. It claims low latency "Handle up to 20000 interrupts per second". So I tried it out and I was able to produce the debounces on the Rpi as well. Here you can find the data recorded with ORM. [2022-09-18_21-01-58_raw.csv](https://github.com/laberning/openrowingmonitor/files/9595247/2022-09-18_21-01-58_raw.csv)

I must say, that library is an extremely interesting find. Can you send me the code how you made it work with ORM on your RPI? I would be very interested in trying this more accurate version of GPOI timer on my machine.

@laberning I wouldn't include it directly in this PR, but it is interesting nonetheless. I think that a potential upgrading of impulse detection (and the handling of the debounce that might be associated with it) would be a separate PR.

> Such noise is not the singe of error but rather the sign of accuracy in my view as it proves that it can handle signals on a higher frequency, hence with better accuracy. I would consider switching to it and implement a one stage doubnce filter. Something like this (note this may not the most efficient code below :)):
> 
> ```js
>      if (hrDelta[0] + hrDelta[1] / 1e9 > config.rowerSettings.minimumTimeBetweenImpulses){
>         hrStartTime = process.hrtime()
>         const delta = hrDelta[0] + hrDelta[1] / 1e9
>         process.send(delta)
>       }
> ```

That could be a solution, but adding this to the configuration could also help according to their documentation:
```js
button.glitchFilter(10000);
```
The latter has the benefit that it will remove short spikes close to the interrupt handler, isolating the rest from ORM from this noise. It will require signals to be at least 10 ms long, effectively systematically delaying detection 10ms across the board (thus not affecting currentDt).

Reply:
@Abasz @JaapvanEkris Totally agreed, this is an area to investigate further, especially for fast spinning rowers this is an issue.

In general Nodejs is not the best runtime environment for real time stuff, that's why currently the GpioTimerService timer runs in a separate process.

Some things that I still have in mind but did not find the time to explore yet are:

- at some point implement an alternative timer process natively with C (actually should be quite straight forward since the interface is pretty simple)
- one of the core design decisions of Open Rowing Monitor is build around the idea, that measurement of time (with the TimerService) is decoupled from the interpretation of it in the engine. One of the advantages of this is, that actually measuring the time is the only thing that has real time requirements and everything else basically works on a stream of dTs (mostly). This would also allow us to build a little piece of hardware for doing the real time measurement and then just send the stream of dTs via serial connection to the Raspi. Something like an optocoupler combined with a little hardware timer.

pigpio looks interesting as well, my guess is that it would hit the same limits that onoff, als benchmarks look similar: https://github.com/fivdi/onoff#benchmarks
Also note that the current onoff has the option to debounce as well that you could try for experiments (see: https://github.com/fivdi/onoff#debouncing-buttons)

@Abasz would be cool if you could move the findings and thoughts that we have collected on this so far into a new issue so we can track progress on this there (and maybe close comments on them here so we don't loose track). Eventually I think this could leave to a PR that we then might integrate into the v1beta branch... ❤️

Reply:
> @Abasz would be cool if you could move the findings and thoughts that we have collected on this so far into a new issue so we can track progress on this there (and maybe close comments on them here so we don't loose track). 

Done.